src/client/options/mod.rs

```diff
@@ -673,41 +680,177 @@ impl Serialize for ClientOptions {
+    ///
+    /// The default value is 10 seconds.
     pub connect_timeout: Option<Duration>,
+
+    /// Whether or not the client should retry a read operation if the operation fails.
+    ///
+    /// The default value is true.
     pub retry_reads: Option<bool>,
+
+    /// Whether or not the client should retry a write operation if the operation fails.
+    ///
+    /// The default value is true.
     pub retry_writes: Option<bool>,
-    pub socket_timeout: Option<Duration>,
-    pub zlib_compression: Option<i32>,
+
+    /// Specifies whether the Client should directly connect to a single host rather than
+    /// autodiscover all servers in the cluster.
+    ///
+    /// The default value is false.
     pub direct_connection: Option<bool>,
+
+    /// The credential to use for authenticating connections made by this client.
     pub credential: Option<Credential>,
+
+    /// Default database for this client.
+    ///
+    /// By default, no default database is specified.
     pub default_database: Option<String>,
-    max_staleness: Option<Duration>,
+
+    /// Whether or not the client is connecting to a MongoDB cluster through a load balancer.
+    pub load_balanced: Option<bool>,
+
+    /// Amount of time spent attempting to send or receive on a socket before timing out; note that
+    /// this only applies to application operations, not SDAM.
+    pub socket_timeout: Option<Duration>,
+
+    /// Specifies the level of compression when using zlib to compress wire protocol messages; -1
+    /// signifies the default level, 0 signifies no compression, 1 signifies the fastest speed, and
+    /// 9 signifies the best compression.
+    pub zlib_compression: Option<i32>,
+
+    /// The database that connections should authenticate against.
+    pub auth_source: Option<String>,
+
+    /// Default read preference for the client.
+    pub read_preference: Option<ReadPreference>,
+
+    wait_queue_timeout: Option<Duration>,
     tls_insecure: Option<bool>,
+    #[cfg(test)]
+    original_uri: String,
+}
+
+/// Elements from the connection string that are not top-level fields in `ConnectionString`.
+#[derive(Default)]
+struct ConnectionStringParts {
+    read_preference_tags: Option<Vec<TagSet>>,
+    max_staleness: Option<Duration>,
     auth_mechanism: Option<AuthMechanism>,
-    auth_source: Option<String>,
     auth_mechanism_properties: Option<Document>,
-    read_preference: Option<ReadPreference>,
-    read_preference_tags: Option<Vec<TagSet>>,
-    load_balanced: Option<bool>,
-    original_uri: String,
+}
+
+/// Specification for mongodb server connections.
+#[derive(Debug, PartialEq)]
+#[non_exhaustive]
+pub enum HostInfo {
+    /// A set of addresses.
+    HostIdentifiers(Vec<ServerAddress>),
+    /// A DNS record for SRV lookup.
+    DnsRecord(String),
+}
+
+impl Default for HostInfo {
+    fn default() -> Self {
+        Self::HostIdentifiers(vec![])
+    }
+}
+
+impl HostInfo {
+    fn into_addresses(self) -> Vec<ServerAddress> {
```

Comment:
Similar to my comment about about the `From` implementation, this kind of discards the distinction between an actual host and a DNS name used for SRV lookup (which contains actual hosts), so I think we should avoid having a method like this.

Reply:
Inlined this into `parse_without_srv_resolution`.